### PR TITLE
Create invoice repository

### LINKED ISSUE #22: Create InvoiceRepository Class
The InvoiceRepository is responsible for holding and searching our Invoice instances. It offers the following methods:

```rb
all - returns an array of all known Invoice instances
find_by_id - returns either nil or an instance of Invoice with a matching ID
find_all_by_customer_id - returns either [] or one or more matches which have a matching customer ID
find_all_by_merchant_id - returns either [] or one or more matches which have a matching merchant ID
find_all_by_status - returns either [] or one or more matches which have a matching status
create(attributes) - create a new Invoice instance with the provided attributes. The new Invoice’s id should be the current highest Invoice id plus 1.
update(id, attribute) - update the Invoice instance with the corresponding id with the provided attributes. Only the invoice’s status can be updated. This method will also change the invoice’s updated_at attribute to the current time.
delete(id) - delete the Invoice instance with the corresponding id
```

The data can be found in data/invoices.csv so the instance is created and used like this:

```rb
se = SalesEngine.from_csv({:invoices => "./data/invoices.csv"})
invoice = se.invoices.find_by_id(6)
# => <invoice>
```

Effort point 3

### PR DESCRIPTION
**Pull request by sami-p**
### Changes
- Created Invoice Repository Class and Spec

### Checklist
- [x] Tests pass
- [x] Ready to merge

closes #22

### CODE REVIEW

File: spec/invoice_repository_spec.rb [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      invoice_repository.create(new_invoice)
+
+      expect(invoice_repository.invoices.length).to eq 11
+
+      invoice_repository.delete(10)
+
+      expect(invoice_repository.invoices.length).to eq 10
+    end
+
+    it 'does not delete anything without a valid id' do
```
Comment (NoahZinter) [inline spec/invoice_repository_spec.rb:181]: Nice, I can tell from this test it's spec harness compliant

File: spec/invoice_spec.rb [hunk 2]
```diff
@@ -88,5 +88,46 @@
...
+                        id: 10,
+                        customer_id: 7,
+                        merchant_id: 8,
+                        status: 'pending',
+                        created_at: Time.now,
+                        updated_at: Time.new(2020, 12, 31)
+                      })
+
+      i.update_time
+      expect(i.updated_at).to be > Time.new(2020, 12, 31)
```
Comment (NoahZinter) [inline spec/invoice_spec.rb:130]: i like this way of testing time updates too.

File: lib/invoice.rb [hunk 3]
```diff
@@ -14,4 +14,16 @@ def initialize(transaction_details)
     @created_at = transaction_details[:created_at]
     @updated_at = transaction_details[:updated_at]
   end
+
+  def update_id(id)
+    @id = id
+  end
+
+  def update_status(status)
+    @status = status unless status.nil?
```
Comment (NoahZinter) [inline lib/invoice.rb:23]: 🗣 Does the spec harness want @status to be a symbol?

**Review by NoahZinter — APPROVED** [review]
I'd double-check whether spec harness wants @status to be a :symbol before merging

File: lib/invoice.rb [hunk 4]
```diff
@@ -14,4 +14,16 @@ def initialize(transaction_details)
     @created_at = transaction_details[:created_at]
     @updated_at = transaction_details[:updated_at]
   end
+
+  def update_id(id)
+    @id = id
```
Comment (redferret) [inline lib/invoice.rb:19]: not sure if this is needed, it's not common to see ids change, unless it's an id linking to another model of data

**Review by redferret — APPROVED** [review]
This looks great!

**Pull request merged**